### DISCUSSION TITLE
What is the best solution to transform data using apiConfigService?

### DISCUSSION THREAD

**Discussion by vietlait** [NarHakobyan/awesome-nest-boilerplate#266] [General]
Hi guy,
With your current structure.
And in my case, `users` table has an avatar column with path ex: `/images/example.jpg.`
When data is responded to, I want to transform it into `https://cdn.my-website.com/images/example.jpg`.
What is the best solution to this problem?
Thanks!

Your answer is to use process.env means using apiConfigService.

But if I want to use it for all user entities. 

I was forced to put it in user.dto class, and in DTO I can't use apiConfigService.

Is there a more optimal solution?

**Comment by duanvnc** [comment]
Hi @vyvian1712 
you can define a function `getFullUrl` in this file to transform your full avatar URL: /src/shared/services/aws-s3.service.ts
```
   getFullUrl(avatar: string): string {
        return `${process.env.CDN_URL}/${avatar}`;
   }

async uploadImage(file: IFile): Promise<string> { ....
```
with env var: CDN_URL=https://cdn.my-website.com

In my case I'm using the AWS S3 to upload.

You can inject AwsS3Service anywhere. Good luck!

**Reply by vietlait** [reply]
Hi @duanvnc,
Thank for your answer.
My question is not how to get the full avatar url.
Which is: how to make a generic transformation for the user response data.
For example for this case I have 2 endpoints to get user details:
1. `auth/me` (Responding to the currently logged in user information)
2. `users/{id}` (Responding to user information by ID parameter)

In the respective controllers, I use `userEntity.toDto()` to return data to the client.
If I use the `getFullUrl` from awsS3Service method, will I have to inject it into the controller?
I don't want that, it looks like what I want would be like this (in user.dto.ts):
```
constructor(user: UserEntity, options?: UserDtoOptions, configService: ApiConfigService) {
    super(user);
    this.firstName = user.firstName;
    this.lastName = user.lastName;
    this.role = user.role;
    this.email = user.email;
    this.avatar = `${configService.cdnBaseURL}/${user.avatar}`;
    ...
  }
```
But injecting `ApiConfigService` here doesn't work.

**Reply by duanvnc** [reply]
just try this simple way in your controller:
```
    @Get(':id')
    @ApiOperation({ summary: 'Get user detail' })
    @ApiOkResponse({ type: ResponseSingle(UserDto) })
    async getUserById(@Param() id: number): Promise<any> {
        const res = await this._service.getUserById(id);
        return this._entityToDto(res);
    }

    private _entityToDto(entity: UserEntity): UserDto {
        if (entity.avatar) {
            entity.avatar = this._awsS3Service.getFullUrl(entity.avatar);
        }
        return new UserDto(entity);
    }
```

the DTO just helps you validate/consist the object where need to use it, shouldn't inject any dynamic variable to there and it will hard to maintain.
